tensorflow/c/experimental/filesystem/plugins/gcs/design-proposal.md

```diff
@@ -0,0 +1,101 @@
+  + When we call `Read` again on that object, first it sends a request to the server to check if the metadata has been changed or not ( by using `expiring_lru_cache<objectmetadata>` ). Next, it calculates how many bytes has been cached. Lastly, it fetches only the missing bytes from server.
+  ```cpp
+  // We want to read from 3 -> 5
+  Read() 
+  // fetches and save to a cached file named fname@3 with size 2
+
+  // Next we read from 1 -> 4
+  Read() 
+  // fetches from 1 -> 3 and save to a cached file name fname@1 with size 2
+  // Next, it copies the content of fname@1 and fname@3 ( from 3 -> 4 ) to the result
+  ```
+- Deps: env, mutex
+- Solution: `env` is used only for getting time. We could pass a callable object that return time instead of `env`. With the mutex, we may have to use `std::mutex` and `std::shared_mutex`
+- Note:
+  + We may want to rewrite `ram_file_block_cache` only, `file_block_cache` is just an interface.
+  + We may want to add `ram_file_block_cache` to a seperate folder `plugins/ram_cache` so any filesystem needs to use cache can use this ( `s3` )
+  + `Status` will be converted to `TF_Status`. Some functions maybe return  `void` and set the `TF_Status*` instead of return `Status`
+
+### gcs_dsn_cache
+DnsCache is a userspace DNS cache specialized for the GCS filesystem
+
+### gcs_throttle
+GcsThrottle is used to ensure fair use of the available GCS capacity
+- Usage: GcsThrottle operates around a concept of tokens. Tokens are consumed when making requests to the GCS service. Tokens are consumed both based on the number of requests made, as well as the bandwidth consumed (response sizes).
+- Note: We don't have much controls about throttle when working with `google-cloud-cpp`. So there are 3 possible solutions:
+  + one call to `gcs::Client` is count as one request
+  + let `gcs::Client` controls the throttle
+  + fork `google-cloud-cpp` and add some controls for throttle
+
+### gcs_file_system
+GcsStatsInterface
+- An object to collect runtime statistics from the GcsFilesystem
+- We may want to simply save those stats to `filesystem->plugin_filesystem`
+- I can't find any implementation of this interface. Maybe we could drop this feature ?
+
+TimeoutConfig
+- Structure containing the information for timeouts related
+-  We don't have much controls about times out so we have to fork `google-cloud-cpp` if we want to use this.
+
+RandomAccessFile
+- If caching is on, we use file_block_cache
+- If caching is off, we read directly to buffer and return to Tensorflow
+
+WritableFile
+- Writing to a temporary file and upload that file to server ( already implemented in Modular plugins )
+
+Directory related operation
+- If the directory is a bucket, we will mimic the current implementation
+- If the directory is an object, I think we could set the `status` to `TF_OK`, do nothing else and return. The current implementation append to the object's name a `/` and treated it like an empty object. It is useless.
```

Comment:
Adding a `/` is the root of several bugs, so let's not do this

Reply:
So we will just set the `TF_OK` or we don't add '/' and treat it as an empty object ?

Reply:
Let's not add the `/` at the end